linkerd/identity/src/imp/rustls.rs

```diff
@@ -0,0 +1,415 @@
+        // don't bother looking at `_acceptable_issuers`.
+        self.resolve_(sigschemes)
+    }
+
+    fn has_certs(&self) -> bool {
+        true
+    }
+}
+
+impl CertResolver {
+    fn resolve_(
+        &self,
+        sigschemes: &[rustls::SignatureScheme],
+    ) -> Option<rustls::sign::CertifiedKey> {
+        if !sigschemes.contains(&SIGNATURE_ALG_RUSTLS_SCHEME) {
+            debug!("signature scheme not supported -> no certificate");
+            return None;
+        }
+        Some(self.0.clone())
+    }
+}
+
+impl rustls::ResolvesServerCert for CertResolver {
+    fn resolve(&self, hello: rustls::ClientHello<'_>) -> Option<rustls::sign::CertifiedKey> {
+        let server_name = if let Some(server_name) = hello.server_name() {
+            server_name
+        } else {
+            debug!("no SNI -> no certificate");
+            return None;
+        };
+
+        // Verify that our certificate is valid for the given SNI name.
+        let c = (&self.0.cert)
+            .first()
+            .map(rustls::Certificate::as_ref)
+            .unwrap_or(&[]); // An empty input will fail to parse.
+        if let Err(err) =
+            webpki::EndEntityCert::from(c).and_then(|c| c.verify_is_valid_for_dns_name(server_name))
+        {
+            debug!(
+                "our certificate is not valid for the SNI name -> no certificate: {:?}",
+                err
+            );
+            return None;
+        }
+
+        self.resolve_(hello.sigschemes())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Crt {
+    pub(crate) id: LocalId,
+    expiry: SystemTime,
+    chain: Vec<rustls::Certificate>,
+}
+
+impl Crt {
+    pub fn new(
+        id: LocalId,
+        leaf: Vec<u8>,
+        intermediates: Vec<Vec<u8>>,
+        expiry: SystemTime,
+    ) -> Self {
+        let mut chain = Vec::with_capacity(intermediates.len() + 1);
+        chain.push(rustls::Certificate(leaf));
+        chain.extend(intermediates.into_iter().map(rustls::Certificate));
+
+        Self { id, chain, expiry }
+    }
+
+    pub fn name(&self) -> &Name {
+        self.id.as_ref()
+    }
+}
+
+#[derive(Clone)]
+pub struct ClientConfig(rustls::ClientConfig);
+
+impl ClientConfig {
+    pub fn empty() -> Self {
+        Self(rustls::ClientConfig::new())
+    }
+
+    pub fn set_protocols(&mut self, protocols: Vec<Vec<u8>>) {
+        self.0.set_protocols(&*protocols)
+    }
+}
+
+impl From<ClientConfig> for Arc<rustls::ClientConfig> {
+    fn from(cc: ClientConfig) -> Arc<rustls::ClientConfig> {
+        Arc::new(cc.into())
+    }
+}
+
+impl From<ClientConfig> for rustls::ClientConfig {
+    fn from(cc: ClientConfig) -> rustls::ClientConfig {
+        cc.0
+    }
+}
+
+impl AsRef<rustls::ClientConfig> for ClientConfig {
+    fn as_ref(&self) -> &rustls::ClientConfig {
+        &self.0
+    }
+}
+
+impl fmt::Debug for ClientConfig {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("")
```

Comment:
i think this should be
```suggestion
        f.debug_struct("ClientConfig")
```